### PR TITLE
fix(config/inherit): resolve presets

### LINKED ISSUE #31278: Inherited config should be resolved before merging
### Describe the proposed change(s).

Discussion: https://github.com/renovatebot/renovate/discussions/31182

To address this:
- We should resolve presets in inherited config prior to merging the config in
- We should log a debug message saying we're doing this
- We should document this behavior clearly, including noting that it's not possible to use an `ignorePresets` entry to ignore a preset in the `extends` of inherited config

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Resolve presets present in inherited config

## Context
- Closes: https://github.com/renovatebot/renovate/issues/31278

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or 
- [x] Both unit tests + ran on a real repository https://github.com/Rahul-renovate-testing/repro-31182/pull/1

### CODE REVIEW

File: docs/usage/config-overview.md
Comment (HonkingGoose) [inline docs/usage/config-overview.md:160]:
```suggestion
If the inherited config has presets, then Renovate:

1. Resolves the presets
1. Adds the resolved presets to the inherited config
1. Merges the presets on top of the global config
```

A numbered list of steps is easier than a long complicated sentence. Please check if this is technically accurate. 😉

File: docs/usage/config-overview.md
Comment (HonkingGoose) [inline docs/usage/config-overview.md:162]:
```suggestion
Using `ignorePresets` in your repository config to ignore certain presets will _not_ affect presets present in the inherited config.
This is because inherited config is resolved _before_ the repository config.
```

This is a difficult sentence. Let's start by splitting it into two smaller parts, and emphasizing the key points.

What are you trying to explain here? I think it is something like this?

> You can _not_ ignore inherited presets with `ignorePresets`, because inherited presets are resolved _before_ the repository config.

File: lib/workers/repository/init/inherited.spec.ts
Comment (HonkingGoose) [inline lib/workers/repository/init/inherited.spec.ts:167]: `Found errors` only tells me something went wrong, but not _what_ went wrong.

Should we print the error message in the log message?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/workers/repository/init/inherited.spec.ts
Comment (RahulGautamSingh) [reply lib/workers/repository/init/inherited.spec.ts:167]: We do: https://github.com/RahulGautamSingh/renovate/blob/140a794f73d908c134c17afcf1845f96fae57440/lib/workers/repository/init/inherited.ts#L121

Comment (HonkingGoose) [reply lib/workers/repository/init/inherited.spec.ts:167]: Thank you for checking! :wink:

I'll let a maintainer resolve my review comment, I can't do that myself...

**Comment by HonkingGoose** [comment]
> ## Context
> 
> Closes:

Do you mean to close an issue with this PR? If yes, you should link to the issue, so that GitHub auto-closes the issue. And so that Semantic Release bot can say which version of Renovate fixed the linked issue.

File: lib/workers/repository/init/inherited.ts
Comment (secustor) [inline lib/workers/repository/init/inherited.ts:118]: You should create a new variable here, for better readability. That way you need no `let` above

File: lib/workers/repository/init/inherited.ts
Comment (secustor) [inline lib/workers/repository/init/inherited.ts:110]:
```suggestion
  if (is.nullOrUndefined(returnConfig.extends)) {
     return mergeChildConfig(config, decryptedConfig);
  }
```
Invert the if clause, to make it more easy to readable

File: lib/workers/repository/init/inherited.ts
Comment (secustor) [inline lib/workers/repository/init/inherited.ts:134]: What happens if there are global configs in the presets? 

We are filtering them out above, but not here.

File: lib/workers/repository/init/inherited.ts
Comment (HonkingGoose) [inline lib/workers/repository/init/inherited.ts:111]:
```suggestion
  // Decrypt before resolving, in case it contains npm authentication for any preset
```

File: lib/workers/repository/init/inherited.ts
Comment (HonkingGoose) [inline lib/workers/repository/init/inherited.ts:114]:
```suggestion
  // Decrypt after resolving, in case the preset contains npm authentication instead
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/config-overview.md
Comment (HonkingGoose) [inline docs/usage/config-overview.md:166]:
```suggestion
You can _not_ use `ignorePresets` in your repository config to ignore presets _within_ inherited config.
```

The current sentence is still hard to understand.

I think my suggestion is better. Please check for technical accuracy before applying! :wink:

File: docs/usage/config-overview.md
Comment (HonkingGoose) [inline docs/usage/config-overview.md:165]:
```suggestion
##### You can not ignore presets from inherited config
```

I think a sub-heading helps separate the "steps that Renovate does" from the explanation why you can not ignore inherited presets.

File: docs/usage/config-overview.md
Comment (HonkingGoose) [inline docs/usage/config-overview.md:165]:
```suggestion

##### You can not ignore presets from inherited config
```

File: lib/workers/repository/init/inherited.ts
Comment (rarkins) [inline lib/workers/repository/init/inherited.ts:111]: Let's skip this logic. We deprecated npm presets long ago

**Review by rarkins — CHANGES_REQUESTED** [review]
Drop support for npm presets in inheritedConfig

File: lib/workers/repository/init/inherited.ts
Comment (rarkins) [inline lib/workers/repository/init/inherited.ts:116]: npm presets have been deprecated for a long time, please remove this logic to simplify things

File: lib/workers/repository/init/inherited.ts
Comment (rarkins) [inline lib/workers/repository/init/inherited.ts:138]: Inherited config does allow some fields which were previously considered "global", such as `onboarding`. Please check that we don't log in this case

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/config-overview.md
Comment (HonkingGoose) [inline docs/usage/config-overview.md:168]:
```suggestion
1. Merge the presets on top of the global config
```

File: docs/usage/config-overview.md
Comment (HonkingGoose) [reply docs/usage/config-overview.md:164]:
```suggestion
If the inherited config contains `extends` presets, then Renovate will:
```

File: docs/usage/config-overview.md
Comment (HonkingGoose) [reply docs/usage/config-overview.md:167]:
```suggestion
1. Add the resolved preset config to the beginning of the inherited config
```

Maybe this is better. If we go with `@rarkins`'s suggestion, we should keep the `the` word. :wink:

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Small additions to `@rarkins`'s suggestions.

**Review by viceice — COMMENTED** [review]
see open conversations, otherwise LGTM

File: docs/usage/config-overview.md
Comment (HonkingGoose) [inline docs/usage/config-overview.md:166]:
```suggestion
1. Resolve the presets
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/workers/repository/init/inherited.ts
Comment (RahulGautamSingh) [reply lib/workers/repository/init/inherited.ts:138]: Even today `onboarding` is not removed as it has `inheritConfigSupport=true`.

**Review by HonkingGoose — APPROVED** [review]

**Review by secustor — DISMISSED** [review]

File: lib/workers/repository/init/inherited.ts
Comment (viceice) [inline lib/workers/repository/init/inherited.ts:108]: do we really need another variable? can't we rename and reuse `filteredConfig`?

File: lib/workers/repository/init/inherited.ts
Comment (viceice) [inline lib/workers/repository/init/inherited.ts:131]:
```suggestion
```
duplicate

File: lib/workers/repository/init/inherited.ts
Comment (RahulGautamSingh) [reply lib/workers/repository/init/inherited.ts:108]: I think using `returnConfig` is more readable and cleaner. Reusing same var might get bit confusing to track.

File: lib/workers/repository/init/inherited.ts
Comment (RahulGautamSingh) [reply lib/workers/repository/init/inherited.ts:131]: `validateConfig` returns both `errors` & `warnings` so it's not a duplicate.

File: lib/workers/repository/init/inherited.ts
Comment (RahulGautamSingh) [reply lib/workers/repository/init/inherited.ts]: On second thought, even having too many vars can be confusing.

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
@rarkins this is blocked by your requested changes

**Review by rarkins — APPROVED** [review]

**Pull request merged**
